channeldb/migration12/invoices.go

```diff
@@ -339,7 +361,73 @@ func deserializeHtlcs(r io.Reader) (map[CircuitKey]*InvoiceHTLC, error) {
 //
 // nolint: dupl
 func serializeInvoice(w io.Writer, i *Invoice) error {
-	panic("unimplemented")
+	creationDateBytes, err := i.CreationDate.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	settleDateBytes, err := i.SettleDate.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	var fb bytes.Buffer
+	err = i.Terms.Features.EncodeBase256(&fb)
+	if err != nil {
+		return err
+	}
+	featureBytes := fb.Bytes()
+
+	preimage := [32]byte(i.Terms.PaymentPreimage)
+	value := uint64(i.Terms.Value)
+	cltvDelta := uint32(i.FinalCltvDelta)
+	expiry := uint64(i.Expiry)
+
+	amtPaid := uint64(i.AmtPaid)
+	state := uint8(i.Terms.State)
+
+	tlvStream, err := tlv.NewStream(
+		// Memo and payreq.
+		tlv.MakePrimitiveRecord(memoType, &i.Memo),
+		tlv.MakePrimitiveRecord(payReqType, &i.PaymentRequest),
+
+		// Add/settle metadata.
+		tlv.MakePrimitiveRecord(createTimeType, &creationDateBytes),
+		tlv.MakePrimitiveRecord(settleTimeType, &settleDateBytes),
+		tlv.MakePrimitiveRecord(addIndexType, &i.AddIndex),
+		tlv.MakePrimitiveRecord(settleIndexType, &i.SettleIndex),
+
+		// Terms.
+		tlv.MakePrimitiveRecord(preimageType, &preimage),
+		tlv.MakePrimitiveRecord(valueType, &value),
+		tlv.MakePrimitiveRecord(cltvDeltaType, &cltvDelta),
+		tlv.MakePrimitiveRecord(expiryType, &expiry),
+		tlv.MakePrimitiveRecord(paymentAddrType, &i.Terms.PaymentAddr),
+		tlv.MakePrimitiveRecord(featuresType, &featureBytes),
+
+		// Invoice state.
+		tlv.MakePrimitiveRecord(invStateType, &state),
+		tlv.MakePrimitiveRecord(amtPaidType, &amtPaid),
+	)
+	if err != nil {
+		return err
+	}
+
+	var b bytes.Buffer
+	if err = tlvStream.Encode(&b); err != nil {
+		return err
+	}
+
+	err = binary.Write(w, byteOrder, uint64(b.Len()))
+	if err != nil {
+		return err
+	}
+
+	if _, err = w.Write(b.Bytes()); err != nil {
+		return err
+	}
+
+	return serializeHtlcs(w, i.Htlcs)
```

Comment:
Why don't we actually make this a tlv field?

Reply:
Then it also wouldn't be necessary to start the serialized htlcs blob with a length.

Reply:
> Then it also wouldn't be necessary to start the serialized htlcs blob with a length.

i think each one would still need a length, no? perhaps you mean the body wouldn't need to be length prefixed?

imo it'd be preferable to serialize each htlc under it's own circuit key instead, then we eliminate length-prefixed tlv streams altogether. do we think that's worth doing now?

Reply:
Hmm, there are indeed multiple variations. But to add least add the htlc tlv blob as is as a tlv record on the invoice level looks more consistent to me.

Reply:
i think for now we should leave them where they are, since for the time being that would _introduce_ a performance penalty in needed to allocate a backing slice for all htlcs. i'd rather not move them preemptively until we know what the workload/bottlenecks are surrounding it.